### PR TITLE
Support for passing additional variables into mongoose

### PR DESCRIPTION
**Pull request by raulriera**
My PR is a: en

💅 Enhancement

Main update on the:

Framework

Fixes https://github.com/strapi/strapi/issues/1179 adds support for passing additional variables into mongoose. By adding a new object into _database.json_, for the additional variables, at this moment only `debug` is supported, which allows mongoose to dump all the queries in the console like this

```
Mongoose: users-permissions_user.findOne({ _id: ObjectId("5aee1e6b7ae7d334d9247e79") }, { fields: {} })
Mongoose: users-permissions_role.find({ _id: { '$in': [ ObjectId("5aedd8b3e235fa32d54c7b63") ] } }, { fields: {} })
Mongoose: appointment.find({ staff: { '$in': [ ObjectId("5aee1e6b7ae7d334d9247e79") ] } }, { fields: {} })
Mongoose: practice.find({ _id: { '$in': [ ObjectId("5aef661dca53d04354a36adc") ] } }, { fields: {} })
Mongoose: users-permissions_permission.findOne({ role: ObjectId("5aedd8b3e235fa32d54c7b63"), type: 'application', controller: 'appointment', action: 'findone', enabled: true }, { fields: {} })
Mongoose: appointment.findOne({ _id: ObjectId("5aefada83ed11a4862ae908a") }, { fields: {} })
Mongoose: patient.find({ _id: { '$in': [ ObjectId("5aee7ee027d7e03b2e0c7c3b") ] } }, { fields: {} })
Mongoose: location.find({ _id: { '$in': [ ObjectId("5af52263900e1c1b1cc2374f") ] } }, { fields: {} })
Mongoose: location.findOne({ _id: ObjectId("5af52263900e1c1b1cc2374f") }, { fields: {} })
Mongoose: practice.findOne({ _id: ObjectId("5aef661dca53d04354a36adc") }, { fields: {} })
[2018-05-14T03:04:00.983Z] debug POST /graphql?_id=5aefada83ed11a4862ae908a (116 ms)
```

### CODE REVIEW

**Comment by raulriera** [comment]
Would love some pointers on how to add tests for this

**Comment by raulriera** [comment]
Hi @lauriejim 

That makes sense, I implemented the changes and it looks something like this

```javascript
instance.connect(uri || `mongodb://${host}:${port}/${database}`, connectOptions);
        
        for (let key in connectOptions) {
          instance.set(key, connectOptions[key])
        }
```

But then I have to use `connectOptions` twice, for the connection and for looping through all of them, should I create a second options internally to handle the "looping" part?

```javascript
// Connect to mongo database
const connectOptions = {}
const options = {} // or something similar
```

**Comment by lauriejim** [comment]
@raulriera I think we have to update `const { authenticationDatabase, ssl } = _.defaults(connection.options, strapi.config.hook.settings.mongoose);`
We have to use an `option` object in default hook value. An put all options in.
And then loop on options only to apply ` instance.set`

**Comment by raulriera** [comment]
Updated, I think this is what you meant right? still very new to this so I don't know what you mean by `default hook value`

File: packages/strapi-mongoose/lib/index.js [hunk 1]
```diff
@@ -69,8 +71,14 @@ module.exports = function (strapi) {
 
         connectOptions.ssl = ssl === true || ssl === 'true';
 
+        options.debug = debug === true || debug === 'true';
```
Comment (kamalbennani) [inline packages/strapi-mongoose/lib/index.js]: Tiny Tip: you can use `Boolean(debug)` in this case which will do exactly what you're doing here

Comment (raulriera) [reply packages/strapi-mongoose/lib/index.js]: Thanks! updated, also changed `ssl` :)

**Comment by raulriera** [comment]
Is there anything missing in this PR?

**Comment by kamalbennani** [comment]
LGTM, I've tested it and it's working :+1:

**Comment by lauriejim** [comment]
Hey @raulriera I would like some changes and I will explain you why.

Use an `options` key to separate database infos and `optional` configs. It's the same as generated `database.json` file.

```js
defaults: {
    defaultConnection: 'default',
    host: 'localhost',
    port: 27017,
    database: 'strapi',
    options: {
      authenticationDatabase: '',
      ssl: false
    }
},
```

Load options configs `const options = _.defaults(connection.options, strapi.config.hook.settings.mongoose.options);` directly in `options` variable (have to remove the empty object declaration).

Then move in `connectOptions` data we need in this object. And remove them from `options` object.

```js
if (!_.isEmpty(options.authenticationDatabase)) {
  connectOptions.authSource = options.authenticationDatabase;
}
connectOptions.ssl = Boolean(options.ssl);

delete options.authenticationDatabase;
delete options.ssl;
```

Now we can add other options in `database.json` than `debug` and it will be handle.
Previously it was limited to `debug`.

Are you agree with this changes ?

**Comment by raulriera** [comment]
Hey @lauriejim, I don't know... A refactor like that seems like out of scope. If we plan to change how the way things work. Probably should be done in a separate PR first. I will close this and just implement them manually in my project

**Comment by lauriejim** [comment]
For only `debug` option your PR is good. Reopen it please to share it with the community 😄

**Pull request merged**